subsys/settings/src/settings_fcb.c

```diff
@@ -79,52 +75,102 @@ int settings_fcb_dst(struct settings_fcb *cf)
 	return 0;
 }
 
-static int settings_fcb_load_cb(struct fcb_entry_ctx *entry_ctx, void *arg)
+/**
+ * @brief Check if there is any duplicate of the current setting
+ *
+ * This function checks if there is any duplicated data further in the buffer.
+ *
+ * @param cf        FCB handler
+ * @param entry_ctx Current entry context
+ * @param name      The name of the current entry
+ *
+ * @retval <0 Negative error code
+ * @retval 0  No duplicates found
+ * @retval >0 Duplicate found
+ */
+static int settings_fcb_check_duplicate(struct settings_fcb *cf,
+					struct fcb_entry_ctx *entry_ctx,
+					const char * const name)
 {
-	struct settings_fcb_load_cb_arg *argp;
-	char buf[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+	struct fcb_entry_ctx entry2_ctx = *entry_ctx;
 	int rc;
 
-	argp = (struct settings_fcb_load_cb_arg *)arg;
-
-	size_t len_read;
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry2_ctx.loc)) == 0) {
+		char name2[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+		size_t name2_len;
 
-	rc = settings_line_name_read(buf, sizeof(buf), &len_read,
-				     (void *)&entry_ctx->loc);
-	if (rc) {
-		return 0;
+		rc = settings_line_name_read(name2, sizeof(name2), &name2_len,
+						     &entry2_ctx);
+		name2[name2_len] = '\0';
+		if (rc) {
+			return rc;
+		}
+		if (!strcmp(name, name2))
+			return 1;
 	}
-	buf[len_read] = '\0';
-
-	/*name, val-read_cb-ctx, val-off*/
-	/* take into account '=' separator after the name */
-	argp->cb(buf, (void *)&entry_ctx->loc, len_read + 1, argp->cb_arg);
-	return 0;
+	if (rc == -ENOTSUP) {
+		rc = 0;
+	}
+	return rc;
 }
 
-static int settings_fcb_load_priv(struct settings_store *cs, line_load_cb cb,
-				  void *cb_arg)
+static int settings_fcb_load_priv(struct settings_store *cs,
+				  line_load_cb cb,
+				  void *cb_arg,
+				  bool filter_duplicates)
 {
 	struct settings_fcb *cf = (struct settings_fcb *)cs;
-	struct settings_fcb_load_cb_arg arg;
+	struct fcb_entry_ctx entry_ctx = {
+		{.fe_sector = NULL, .fe_elem_off = 0},
+		.fap = cf->cf_fcb.fap
+	};
 	int rc;
 
-	arg.cb = cb;
-	arg.cb_arg = cb_arg;
-	rc = fcb_walk(&cf->cf_fcb, 0, settings_fcb_load_cb, &arg);
-	if (rc) {
-		return -EINVAL;
+	while ((rc = fcb_getnext(&cf->cf_fcb, &entry_ctx.loc)) == 0) {
+		char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+		size_t name_len;
+		int rc;
+		bool pass_entry = true;
+
+		rc = settings_line_name_read(name, sizeof(name), &name_len,
+					     (void *)&entry_ctx);
+		if (rc) {
+			LOG_ERR("Failed to load line name: %d", rc);
+			return 0;
+		}
+		name[name_len] = '\0';
+
+		if (filter_duplicates) {
+			rc = settings_fcb_check_duplicate(cf, &entry_ctx, name);
+			if (rc < 0)
```

Comment:
shouldn't this be if (rc <= 0) ?

Reply:
No. If 0 is returned I wish to set pass_entry to 1 and continue the loop.

Reply:
There is something strange with the logic here, if no duplicates are found or if a duplicate is found you want to set pass_entry to 1 (unless the description of settings_fcb_check_duplicate is wrong).

Reply:
```pass_entry = (rc == 0);``` - bracket added. If duplicates are found then rc = 1. If no duplicates is found - rc = 0.

Reply:
Ok, this makes it more clear, but I still think it is not correct. Whenever an error is occurring during the find duplicates there is a jump out of the loop, while it should just be a continue. It should still load whatever it has found.

Reply:
Done